docs/posts/flink-unified-stream-batch/index.qmd

```diff
@@ -0,0 +1,414 @@
+[Alibaba](https://www.ververica.com/blog/apache-flinks-stream-batch-unification-powers-alibabas-11.11-in-2020),
+and [Lyft](https://beam.apache.org/case-studies/lyft/) have also adopted
+similar solutions.
+
+Secondly, in the world of machine learning, it’s common for data scientists to
+develop locally and experiment with a sampled, batch dataset in Python. If the
+results look promising, the features and models would then be deployed into
+production. Oftentimes, there is a code handover in this process, and a
+dedicated team of developers would be responsible for rewriting the logic for
+production, as a streaming workload.
+
+In both cases, there is a huge amount of technical overhead. If there is a
+streamlined architecture, using a unified API, much of this overhead can be
+avoided. As a platform engineer, you no longer need to worry about maintaining
+two separate architectures and codebases. As a data scientist or a machine
+learning engineer, you can write one single workload that can execute both on
+batch and streaming backends. Wouldn’t that be amazing?
+
+## Ibis unifies batch and streaming
+
+Enter Ibis. Ibis unifies batch and streaming with a single API. It decouples
+the dataframe API from backend execution, so that the logic for defining data
+transformations is unaffected by implementation discrepancies across backend
+engines. There is also an ongoing effort to further increase interoperability
+across different languages and systems via a standard query plan intermediate
+representation (IR), using a library called
+[`Substrait`](https://substrait.io/).
+
+What does this actually look like? For example, Ibis allows users to define
+window aggregations using the [`over()`
+method](https://ibis-project.org/reference/expression-tables#ibis.expr.types.groupby.GroupedTable.over).
+When executed on the Flink backend, this translates into [Flink’s over
+aggregation query](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/table/sql/queries/over-agg/)
+and outputs an aggregated value for every input row over a range of ordered
+rows. On streaming data, aggregation results are continuously computed and
+written into data sinks (e.g., Kafka, Redis) as records are received at and
+consumed from the upstream data source (e.g., Kafka, Change Data Capture). In
+pandas, the conceptual analog is [windowing
+operation](https://pandas.pydata.org/docs/user_guide/window.html). Results are
+computed by looking back the length of the window from the current observation,
+but can be computed all at once because batch data is static.
+
+Another great example is deduplication. In Flink SQL, this looks something like this:
+
+```sql
+SELECT [column_list]
+FROM (
+   SELECT [column_list],
+     ROW_NUMBER() OVER ([PARTITION BY col1[, col2...]]
+       ORDER BY time_attr [asc|desc]) AS rownum
+   FROM table_name)
+WHERE rownum = 1
+```
+In a database like Postgres, this could be as simple as
+```sql
```

Comment:
would add newline here (and a few places below) for consistency -- doesn't really matter for rendering